File: 0207-course-schedule/0207-course-schedule.java
class Solution {
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        int[] inDegree = new int[numCourses];
        Boolean[] visited = new Boolean[numCourses];
        ArrayList<Integer>[] adjMatrix = new ArrayList[numCourses];
        for(int i = 0; i < numCourses; i++) {
            adjMatrix[i] = new ArrayList<>();
            visited[i] = false;
        }

        for(int[] prerequisite: prerequisites) {
            int v = prerequisite[0];
            int u = prerequisite[1];
            inDegree[v]++;
            adjMatrix[u].add(v);
        }

        Queue<Integer> q = new LinkedList();
        for(int i = 0; i < numCourses; i++) {
            if(inDegree[i] == 0) {
                q.add(i);
                visited[i] = true;
            }
        }

        while(!q.isEmpty()) {
            int u = q.poll();
            for(Integer v: adjMatrix[u]) {
                if(visited[v]) {
                    return false;
                }
                if(--inDegree[v] == 0) {
                    visited[v] = true;
                    q.add(v);
                }
            }
        }
        for(int i = 0; i < numCourses; i++) {
            if(!visited[i])
                return false;
        }
        return true;
    }
}

File: 0207-course-schedule/0207-course-schedule.py
from collections import defaultdict
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        inDegree = [0] * numCourses
        adjMap =  defaultdict(list)
        for course, preCourse in prerequisites:
            inDegree[course] += 1
            adjMap[preCourse].append(course)
        
        def getLeafCourse():
            res = []
            for course, degree in enumerate(inDegree):
                if degree == 0:
                    res.append(course)
            return res

        courses = getLeafCourse()
        totalCourse = len(courses)
        queue = deque(courses)
        while len(queue) > 0:
            curr = queue.popleft()
            inDegree[curr] -= 1
            for course in adjMap[curr]:
                inDegree[course] -= 1
                if inDegree[course] == 0:
                    queue.append(course)
                    totalCourse += 1
            courses = getLeafCourse()
        return totalCourse == numCourses